src/addons/wiiext.cpp

```diff
@@ -30,190 +32,501 @@ void WiiExtensionInput::setup() {
-
-    if (buttonA) gamepad->state.buttons |= GAMEPAD_MASK_B2;
-    if (buttonB) gamepad->state.buttons |= GAMEPAD_MASK_B1;
-    if (buttonX) gamepad->state.buttons |= GAMEPAD_MASK_B4;
-    if (buttonY) gamepad->state.buttons |= GAMEPAD_MASK_B3;
-    if (buttonL) gamepad->state.buttons |= GAMEPAD_MASK_L1;
-    if (buttonZL) gamepad->state.buttons |= GAMEPAD_MASK_L2;
-    if (buttonR) gamepad->state.buttons |= GAMEPAD_MASK_R1;
-    if (buttonZR) gamepad->state.buttons |= GAMEPAD_MASK_R2;
-    if (buttonSelect) gamepad->state.buttons |= GAMEPAD_MASK_S1;
-    if (buttonStart) gamepad->state.buttons |= GAMEPAD_MASK_S2;
-    if (buttonHome) gamepad->state.buttons |= GAMEPAD_MASK_A1;
-    if (dpadUp) gamepad->state.dpad |= GAMEPAD_MASK_UP;
-    if (dpadDown) gamepad->state.dpad |= GAMEPAD_MASK_DOWN;
-    if (dpadLeft) gamepad->state.dpad |= GAMEPAD_MASK_LEFT;
-    if (dpadRight) gamepad->state.dpad |= GAMEPAD_MASK_RIGHT;
+    for (auto currAxis = axesOfChange.begin(); currAxis != axesOfChange.end(); ++currAxis) {
+        if (!currAxis->second.empty()) {
+            axisType = currAxis->first;
+
+            switch (axisType) {
+                case WII_ANALOG_TYPE_LEFT_STICK_X:
+                    gamepad->state.lx = getDelta(currAxis->second, GAMEPAD_JOYSTICK_MID);
+                    break;
+                case WII_ANALOG_TYPE_LEFT_STICK_Y:
+                    gamepad->state.ly = getDelta(currAxis->second, GAMEPAD_JOYSTICK_MID);
+                    break;
+                case WII_ANALOG_TYPE_RIGHT_STICK_X:
+                    gamepad->state.rx = getDelta(currAxis->second, GAMEPAD_JOYSTICK_MID);
+                    break;
+                case WII_ANALOG_TYPE_RIGHT_STICK_Y:
+                    gamepad->state.ry = getDelta(currAxis->second, GAMEPAD_JOYSTICK_MID);
+                    break;
+                case WII_ANALOG_TYPE_LEFT_TRIGGER:
+                    gamepad->state.lt = getDelta(currAxis->second, GAMEPAD_TRIGGER_MID);
+                    break;
+                case WII_ANALOG_TYPE_RIGHT_TRIGGER:
+                    gamepad->state.rt = getDelta(currAxis->second, GAMEPAD_TRIGGER_MID);
+                    break;
+            }
+        }       
+    }
 }
 
-uint16_t WiiExtensionInput::map(uint16_t x, uint16_t in_min, uint16_t in_max, uint16_t out_min, uint16_t out_max) {
-    return (x - in_min) * (out_max - out_min) / (in_max - in_min) + out_min;
+uint16_t WiiExtensionInput::getAverage(std::vector<WiiAnalogChange> const& changes) {
+    std::vector<uint16_t> values;
+
+    if (changes.empty()) {
+        return 0;
+    }
+
+    for (auto currVal = changes.begin(); currVal != changes.end(); ++currVal) {
+        values.push_back(currVal->analogValue);
+    }
+
+    return std::accumulate(values.begin(), values.end(), 0) / values.size();
 }
+
+uint16_t WiiExtensionInput::getDelta(std::vector<uint16_t> const& changes, uint16_t baseValue) {
+    uint16_t value = baseValue;
```

Comment:
Yeah getAverage could do this I think?

Reply:
getAverage is for smoothing out values, where getDelta is for knowing how much of a change needs to be applied to a given point. It adjusts to account for neutral directions resetting the analog state if it changed earlier in the loop.

Example: LS and RS are mapped to L analog output.  LS gets read first, then RS. LS moves, RS stays neutral. 
 Without getDelta, LS adjusts L.X to xFFFF, but RS.X is 7FFF, so L.X stays x7FFF, undoing the change.  With getDelta, the previous change gets adjusted by later changes before sending them out.  This could use some tuning, but it was to prevent later analog neutral states from dropping earlier changes.